src/Simplify_Stmts.cpp

```diff
@@ -218,23 +218,36 @@ Stmt Simplify::visit(const For *op) {
                                           op->for_type == ForType::Vectorized));
 
     bool bounds_tracked = false;
-    if (min_bounds.min_defined || (min_bounds.max_defined && extent_bounds.max_defined)) {
-        min_bounds.max += extent_bounds.max - 1;
-        min_bounds.max_defined &= extent_bounds.max_defined;
-        min_bounds.alignment = ModulusRemainder{};
+    ExprInfo loop_var_info;
+    // Deduce bounds for the loop var that are true for any code than runs
+    // inside the loop body. Code in the inner loop only runs if the extent is
+    // at least one, so we can throw a max around the extent bounds.
+
+    loop_var_info.bounds =
+        ConstantInterval::make_union(min_info.bounds,
+                                     min_info.bounds + max(extent_info.bounds, 1) - 1);
+
+    if (loop_var_info.bounds.max_defined ||
+        loop_var_info.bounds.min_defined) {
         bounds_tracked = true;
-        bounds_and_alignment_info.push(op->name, min_bounds);
+        bounds_and_alignment_info.push(op->name, loop_var_info);
+    }
+
+    Expr extent_positive = mutate(0 < new_extent, nullptr);
+    if (is_const_zero(extent_positive)) {
+        return Evaluate::make(0);
     }
 
     Stmt new_body;
     {
         // If we're in the loop, the extent must be greater than 0.
-        ScopedFact fact = scoped_truth(0 < new_extent);
+        ScopedFact fact = scoped_truth(extent_positive);
         new_body = mutate(op->body);
     }
     if (in_unreachable) {
-        if (extent_bounds.min_defined && extent_bounds.min >= 1) {
-            // If we know the loop executes once, the code that runs this loop is unreachable.
+        if (extent_info.bounds > 0) {
+            // If we know the loop executes at least once, the code that runs
+            // this loop is unreachable.
```

Comment:
Not sure if I'm just struggling to parse this, but this comment doesn't make sense to me (the before or edit). If the lop executes, then the code inside the loop is certainly *reachable*, no?

Reply:
Please take another look and tell me if it's clearer now. I also found a bug here related to not popping the expr info for the loop var, so the code has been reshuffled a little.

Reply:
Change lgtm! Thanks